### ISSUE TITLE
question: increase visibility of SNPs at wider views?

### ISSUE THREAD

**Issue by kevfengler227** [igvteam/igv#1551]
Is there a way to increase the visibility of SNPs at wider views? In the example below, I can see SNP differences between alignments in a 56 kb window, but not in a 140 kb window, which encompasses the entire region I want to display.

![image](https://github.com/user-attachments/assets/71f80b19-ac1a-4730-be58-93fec8a9a784)

![image](https://github.com/user-attachments/assets/c9e5bbb2-2a0c-4c20-a30d-cb5118a33718)

**Comment by kevfengler227** [comment]
I should add the coverage track has the desired SNP visibility, just not the alignments.

![image](https://github.com/user-attachments/assets/4229c33a-0a45-47a7-b2c3-c0af22e60520)

**Comment by jrobinso** [comment]
Probably not at the moment, but I will look into it.    When zoomed in all mismatches are shown, not just those deemed significant for the coverage track.  My vague recollection is we stop doing this at some resolution as it becomes too cluttered, but this should be revisited.

**Comment by jrobinso** [comment]
Just a note here -- this only seems to happen with long read (3rd gen) data.

**Comment by kevfengler227** [comment]
Indeed. These are actually 140 kb genomic segments aligned as HiFi reads. But I am trying to show the SNP variation and haplotypes in each genome. This is one way to turn IGV into a pangenome viewer!

**Comment by jrobinso** [comment]
Interesting.   I'll have this fixed soon.   If the dataset you are using or creating is public let me know, it would be an interesting test case to add.

**Comment by jrobinso** [comment]
One issue that arises as you zoom out is many bases land on the same pixel.    At 100kb approximately 100 bases / pixel.   For typical reads this means that nearly every pixel of the alignment will have a mismatch,  often multiple mismatches.     We might need some user options for how to handle this.

**Comment by jrobinso** [comment]
As illustration here is a 247kb window of pacbio alignments with mismatches drawn.  Its not usable, and rendering is extremely slow.   So some preferences or special mode is needed here
<img width="1233" alt="Screenshot 2024-08-28 at 11 01 21 PM" src="https://github.com/user-attachments/assets/279c6d71-4ea5-40fc-9ba3-56b6de123654">

**Comment by kevfengler227** [comment]
yes, I did not intend to use this capability for PacBio reads or ONT reads, but rather genomes with relatively few differences. So a "genome" mode would be ideal. I can send you an example public dataset.

Of course, the user needs to do some upfront work to create the ideal input data, but reducing each genome to 1x is extremely powerful, rather than 30x PacBio, and the only practical way to view a large pangenome.

**Comment by jrobinso** [comment]
A public dataset would be helpful.

**Comment by jrobinso** [comment]
There will still be limits on zoom out as at a minimum the sequence for the entire region needs to be loaded, not to mention the read sequence in every alignment.   We could not view an entire chromosome with read sequences for example.

**Comment by kevfengler227** [comment]
Admittedly, this will probably only work well for low diversity applications like my initial request. In that case there are only ~13 SNPs in a handful of genomes in a 140 kb range, which was just out of visibility limit,  so I was hoping for way to crank up the SNP visibility, but that wouldn't make sense if there was a ton of variation- which is often the case for plant pangenomes.

It seems that 114 kb is the visibility max for SNPs, but INDELs are visible at much wider ranges.

![image](https://github.com/user-attachments/assets/de1f6c9a-d09a-4929-8768-61478e8576f8)

![image](https://github.com/user-attachments/assets/a83abb1d-7ba5-4a98-9d95-8b211345f1e8)

But this real world example from the maize pangenome probably has too many SNPs to display nicely at wider-ranges, but in some specific cases it would still be useful

**Comment by kevfengler227** [comment]
here genomes were aligned in 100 kb consecutive chunks

**Comment by kevfengler227** [comment]
Here is a test dataset of mock data, with a few SNPs over 245 kb

![image](https://github.com/user-attachments/assets/edcc8559-7ab1-473e-a046-deb24599ddb2)
[10genomes.fasta.gz](https://github.com/user-attachments/files/16805218/10genomes.fasta.gz)

[test.fasta.gz](https://github.com/user-attachments/files/16805219/test.fasta.gz)

**Comment by kevfengler227** [comment]
minimap2 -ax map-hifi -s 1000 -t40 test.fasta 10genomes.fasta | samtools view  -@20 -b -1 - | samtools sort -@20 --write-index  -o 10genomes.bam

**Comment by kevfengler227** [comment]
So basically trying to use  IGV has a haplotype-viewer

**Comment by jrobinso** [comment]
I've never used minimap2 but that's o.k.   I think the simplest resolution of this issue would be to just make the max window for showing mismatches user settable,  probably as a preference.   A new display mode is a bigger topic that deserves its own issue, and would be longer term and prioritized vs other bigger topics.

I will also make snp display subject to the limit.  BTW currently the limit is not on the genomic window, which can vary by display size, but on the resolution in bp / pixel

**Comment by kevfengler227** [comment]
sounds great. thanks!

**Comment by baozg** [comment]
Related question: If loading IGV with more than 100 genomes (wholge genome alignment by minimap2 -x asm20), the speed would be very slow. If there any way to speed it up?

**Comment by kevfengler227** [comment]
Rather then performing whole-genome alignments, I typically align consecutively 10kb chunked genomes, which is faster for alignment and the alignments can be toggled by mapping quality and alignment score. If you add the genome name to the read group when running minimap2 and merge the resulting bam files, 100 genomes is essentially the same as 100x Illumina coverage and is quite rapid to view in IGV. 

![image](https://github.com/user-attachments/assets/56519743-83fd-454f-8306-a08a21c97a14)

**Comment by kevfengler227** [comment]
If you zoom out you can see the PAV in the variations well, just not the SNPs

![image](https://github.com/user-attachments/assets/6633627d-0708-4e6a-9c77-9a367071c712)

**Comment by kevfengler227** [comment]
coloring and grouping by read group is key

**Comment by kevfengler227** [comment]
finally, if you number the chunks consecutively you know exactly where is came from in the query- which is much better than using kmers or other methods where coordinates are lost. Then you know you are looking at syntenic alignments when you mouse over a chunk and see it's chunk# (position) is similar to reference

**Comment by baozg** [comment]
Thanks for sharing! Chunking could be a good idea, but this also lose the abiltiy to detect the variation longer than chunk length or introduce ambiguous alignment (TEs). It more like chain by yourself as you know the coordinates. I think it would be better if IGV use chunk in the browser but with more contiguous alignments. Actually, I use AnchorWave and wfmash more often, whihc nearly produce end-to-end alignment in A.thaliana (easier than maize). For the alternative approach other than IGV, I use https://github.com/cmdcolin/jbrowse-plugin-mafviewer for convert my paf to pseduomaf (which only can present SNPs or DEL)

**Comment by kevfengler227** [comment]
you can use whatever chunk size you want for a given application depending on the level of similarity in pangenome, typically 1-100 kb (aligned with map-hifi). With that you can see quite large INDELs. Again, you can control what is displayed by changing the visualization parameters in IGV more so than with whole-genome alignments. Also, the directionality of chunks is indicative of inversions. For major differences the lack of an aligned chunk also informative.

so the real beauty of the chunked alignment approach is that is highly parallelizable and rapid. One can do an all-by-all comparison in minutes, so that all/any reference(s) can be viewed in IGV with all queries on a whim. If you want to get fancy you can group your queries into various sub-groups, rather than 1 big one.
